### DISCUSSION TITLE
How to build `miniopenvpn` used in `debug-openvpn.sh`?

### DISCUSSION THREAD

**Discussion by scruel** [schwabe/ics-openvpn#1719] [General]
Hi, I want to build openvpn binaries on Android, however I couldn't find anything related in both [OpenVPN3 official repo](https://github.com/OpenVPN/openvpn3) or here.

I can only find related usages about `miniopenvpn`, but not build instructions:
https://github.com/schwabe/ics-openvpn/blob/4473ad1f1bc49c166fb8ab92ee7a7b91c608135e/main/misc/prepareovpn3.sh
https://github.com/schwabe/ics-openvpn/blob/4473ad1f1bc49c166fb8ab92ee7a7b91c608135e/misc/debug-openvpn.sh

Couldn't find `miniopenvpn` on my phone:
```shell
/data/data/de.blinkt.openvpn/cache $ ls
logcache.dat  mgmtsocket
```
And related file is just empty:
https://github.com/schwabe/ics-openvpn/blob/4473ad1f1bc49c166fb8ab92ee7a7b91c608135e/main/src/main/cpp/minivpn/minivpn.c

@schwabe All related android build issues are pointed to your repo by yourself, so hope you can help a little bit more! Thanks!

**Comment by schwabe** [comment]
I have not used that debug-openvpn.sh script in years. It used to work but I don't think it still works. Current binaries are build using cmake and using the CmakeLists.txt in the cpp directolry.

**Reply by scruel** [reply]
https://github.com/schwabe/ics-openvpn/tree/011bcb7922f5e23a04322f91318704b1e5ab231d
Oh, 12 years ago

**Comment by scruel** [comment]


**Reply by schwabe** [reply]
I never tried to build it outside Android studio/gradlew build. You can also take a look at the github actions (https://github.com/schwabe/ics-openvpn/blob/master/.github/workflows/build.yaml) to get an idea how to build using command line. It is probably possible to build the code outside Android studio/gradlew but the current cmake file is probably not very useful as it has building OpenVPN with TARGET_ANDROID hardcoded. And TARGET_ANDROID pretty much requires OpenVPN to run with an app like mine to work as it effectively puts all network configuration into the management interface

**Comment by scruel** [comment]
Hi again, I've built it successfully:
```
$ ./pie_openvpn.arm64-v8a  --version
OpenVPN 2.7-icsopenvpn [git:icsopenvpn/v0.7.51-0-g0ad44976] arm64-v8a [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on May 15 2024
library versions: OpenSSL 3.2.0 23 Nov 2023, LZO 2.10
Originally developed by James Yonan
Copyright (C) 2002-2023 OpenVPN Inc <sales@openvpn.net>
```
But how can I switch the core to openvpn3 rather than openvpn2?
